community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -142,7 +143,7 @@ TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKerne
     private TransactionHooks.TransactionHooksState hooksState;
     private final KernelStatement currentStatement;
     private final StorageStatement storageStatement;
-    private CloseListener closeListener;
+    private final List<CloseListener> closeListeners = new ArrayList<>( 2 );
```

Comment:
Fine, if you dislike the original design, we can change it. But I don't think this PR should be blamed for the current suboptimal design and also I don't think this PR is the right place for that refactor.